DietPi-Drive_Manager | Add NFSv3 handling

MichaIng opened the pull request:
**Status**: Ready
- [x] Changelog entry

Hijack:
- [x] Remove `ntfs-3g` and `hfsplus` from DietPi core requirements
- [x] Remove them on patch, if no related drive is detected.
- [x] Remove `initramfs-tools` from RPi images, since it is not required for RPi bootloader, add "kmod" and "udev" to list instead, which would be autoremoved otherwise.
  - https://github.com/Fourdee/DietPi/issues/1858#issuecomment-401500992
- [x] Changelog entry

**Reference**: https://github.com/Fourdee/DietPi/issues/1940

**Commit list/description**:
+ DietPi-Drive_Manager | Add NFSv3 handling: Catch error that indicates NFSv3, then allow required server-side path to share input.
  - 🈯️ Jessie, Stretch, Buster
+ DietPi-Drive_Manager | Minor coding
+ CHANGELOG | DietPi-Drive_Manager: Added NFSv3 mount support

Hijack:
+ DietPi-Drive_Manager | Fix "tail: cannot open ‘+2’ for reading: No such file or directory"
  - `tail -n +2` was fix, but `lsblk -n` skips headings output, thus is the better solution
+ DietPi-Drive_Manager | Check and install APT packages, required to read all detected file systems
  - This allows images to be shipped without `ntfs-3g` and `hfsplus`. Most users will not require them and in case, drive manager now installs them as fast as related file systems are detected.
  - 🈯️ Test succesed
+ DietPi-PREP | Remove "ntfs-3g" and "hfsplus" from images, which will be auto installed by dietpi-drive_manager ondemand instead.
+ DietPi-PREP | Remove "initramfs-tools" from RPi images, which is not required for their bootloader.
+ DietPi-Patchfile | Remove "ntfs-3g" respectively "hfsplus", if no attached drive requires it
+ DietPi-Patchfile | "apt-mark auto" all packages that we aim to purge, then autoremove with a single final G_AGA
  - This speeds up patch and keeps dependencies of manual installed packages.
+ CHANGELOG | DietPi-Drive_Manager: Now installs packages automatically, that are required to read all attached drive file systems
+ CHANGELOG | RPi: New RPi images will be shipped without "initramfs-tools" preinstalled.
+ DietPi-PREP | Add "kmod" and "udev" to core requirements, which are dependencies of "initramfs-tools", which is not installed on RPi any more.
+ DietPi-PREP | Purge "os-prober" again after execution
+ DietPi-Patchfile | To be failsafe, use "G_AGP" instead of "apt-mark auto" for packages, which features can actively collide with DietPi controls, currently: "dphys-swapfile", "ntp"
+ DietPi-Drive_Manager | Fix "/mnt/" accumulation within $nfs_fp_mount_target with every loop and in case, user entered "/mnt/*" as mount target.
+ DietPi-Drive_Manager | Minor coding and wording
+ Moved all RootFS R/W check code into dietpi-globals, to vastly speed up checks and avoid doubled/parallel drive_manager execution.

MichaIng committed:
v6.13

+ DietPi-Drive_Manager | Add NFSv3 handling: Catch error that indicates NFSv3, then allow required server-side path to share input.
+ DietPi-Drive_Manager | Minor coding

MichaIng committed:
v6.13

+ CHANGELOG | DietPi-Drive_Manager: Added NFSv3 mount support

MichaIng committed:
v6.13

+ DietPi-Drive_Manager | Fix "tail: cannot open ‘+2’ for reading: No such file or directory"
+ DietPi-Drive_Manager | Check and install APT packages, required to read all detected file systems

MichaIng committed:
v6.13

+ DietPi-PREP | Remove "ntfs-3g" and "hfsplus" from images, which will be auto installed by dietpi-drive_manager ondemand instead.
+ DietPi-PREP | Remove "initramfs-tools" from RPi images, which is not required for their bootloader.

MichaIng committed:
v6.13

+ DietPi-Patchfile | Remove "ntfs-3g" respectively "hfsplus", if no attached drive requires it
+ DietPi-Patchfile | "apt-mark auto" all packages that we aim to purge, then autoremove with a single final G_AGA

MichaIng committed:
v6.13

+ DietPi-Drive_Manager | Now installs packages automatically, that are required to read all attached drive file systems
+ Image | RPi: New RPi images will be shipped without "initramfs-tools" preinstalled.

MichaIng committed:
v6.13

+ DietPi-PREP | Add "kmod" and "udev" to core requirements, which are dependencies of "initramfs-tools", which is not installed on RPi any more.
+ DietPi-PREP | Purge "os-prober" again after execution

MichaIng committed:
v6.13

+ DietPi-Patchfile | Syntax fix

MichaIng committed:
v6.13

+ CHANGELOG | "ntfs-3g" and "hfsplus" are needed to write to those file systems, not read

MichaIng committed:
v6.13

+ DietPi-Drive_Manager | "ntfs-3g" and "hfsplus" are required for write access

MichaIng commented:
Just found, mounting `ntfs` creates a background process (at least on Jessie), looks like `mount.ntfs` by default is no oneshot there. Killing the process removes the mount.

Never saw that. Or is it related to systemd USB automount somehow? `mnt-usb.automount`

Fourdee commented on dietpi/dietpi-drive_manager:
```diff
@@ -490,6 +501,13 @@ aDRIVE_ISPARTITIONTABLE ${aDRIVE_ISPARTITIONTABLE[$i]}
+		# Install needed file system write packages
+		if [[ $need_ntfs || $need_hfs ]]; then
+
+			G_AG_CHECK_INSTALL_PREREQ $need_ntfs $need_hfs
```
Minor, these could probably be converted to 1 var?
```
pre_req_packages=''
pre_req_packages+='ntfs3g'
pre_req_packages+='htfs'
G_AG_CHECK_INSTALL_PREREQ $pre_req_packages
```

MichaIng replied:
@Fourdee 
Thought about the same first, but duplicates would be possible, as we loop though all drives. Would be needed to scan for already added strings and the overhead of this is larger than simply two variables, I though.

But I guess with some coffee, we could find an elegant way with one variable and no additional grep calls 😆.

Fourdee replied:
@MichaIng 

> But I guess with some coffee, we could find an elegant way with one variable and no additional grep calls 

lol 👍

Fourdee commented on dietpi/patch_file:
```diff
@@ -426,7 +426,7 @@ _EOF_
 			/DietPi/dietpi/func/dietpi-set_dphys-swapfile 1
 
-			G_AGP dphys-swapfile
+			apt-mark auto dphys-swapfile
```
I believe this should be a forced removal, to prevent it from remaining (in rare case?) installed and conflicting with our swapfile system?

MichaIng replied:
@Fourdee 
I checked the priority of all packages and only use auto mark for optional, extra etc., which includes dphys-swapfile. But yeah, not sure if this can vary on systems, properly better G_AGP to be on secure side.

MichaIng replied:
Done: https://github.com/Fourdee/DietPi/pull/1994/commits/b60e8fbd45b770edfa9eacd8a11f9e93235ee436

Fourdee commented on dietpi/patch_file:
```diff
@@ -631,7 +631,7 @@ _EOF_
 				killall -w ntpd &> /dev/null
 
-				G_AGP ntp
+				apt-mark auto ntp
```
Same with this, forced removal?

Fourdee reviewed (changes requested):
Some minor items need review?

Fourdee commented:
@MichaIng 

> Just found, mounting ntfs creates a background process (at least on Jessie)

Yep, same for stretch, although ```umount``` should also kill the process if I remember correctly?
![image](https://user-images.githubusercontent.com/7077026/43706075-47040ade-995c-11e8-9863-7d4cf8264eae.png)

MichaIng commented:
@Fourdee 
Okay, good to know, this seems to be needed for NTFS r/w access. Never recognized this.
Jep, umount stops process and killing process removes mount.

MichaIng committed:
v6.13

+ DietPi-Patchfile | To be failsafe, use "G_AGP" instead of "apt-mark auto" for packages, which features can actively collide with DietPi controls, currently: "dphys-swapfile", "ntp"

MichaIng committed:
v6.13

+ CHANGELOG | Wording

MichaIng committed:
v6.13

+ DietPi-Drive_Manager | NFSv3 support: Verbose mode is needed to catch required mount output.
+ DietPi-Drive_Manager | Remove mount output file as well on cancel/cleanup.

MichaIng committed:
v6.13

+ DietPi-Drive_Manager | Suppress "not found" error messages on tmp file removal. Those rarely occur on user cancel and interruption.

MichaIng committed:
Merge branch 'testing' into NFSv3_handling

MichaIng committed:
v6.13

+ DietPi-Drive_Manager | Fix "/mnt/" accumulation within $nfs_fp_mount_target with every loop and in case, user entered "/mnt/*" as mount target.
+ DietPi-Drive_Manager | Minor coding and wording

MichaIng commented:
Fixed another issue with "/mnt/" adding up several times to nfs mount target string, when loop is done or user entered it into input box. We recognized this already before v6.12, if I remember correctly, but forgot to fix it 😉.

Ant $nfs_clientname is now $nfs_server_ip, don't know if the first is remain from older code, but doesn't fit here.

As well added some example values that mostly should safe user some input time 😄.
_____________
+ Moved all RootFS R/W check code into dietpi-globals, to vastly speed up checks and avoid doubled/parallel drive_manager execution.
  - Fixes: https://github.com/Fourdee/DietPi/issues/1997

MichaIng committed:
v6.13

+ DietPi-Globals | Move RootFS R/W check code into globals, to avoid loading DietPi-Drive_Manager on every call, possibly in parallel

MichaIng committed:
v6.13

+ DietPi-Drive_Manager | Removed RootFS R/W check, which is now fully moved to dietpi-globals

MichaIng committed:
v6.13

+ DietPi-Globals | G_CHECK_ROOTFS_RW(): Doubled speed via grep on "/proc/mounts" instead of "mount" command.

MichaIng committed:
Merge branch 'testing' into NFSv3_handling

MichaIng committed:
v6.13

+ alignment

MichaIng commented:
@Fourdee 
What do you think, ready for merge? I will add new PR about the parallel dietpi-* execution checks.

Fourdee committed:
Merge branch 'testing' into NFSv3_handling

Fourdee commented:
@MichaIng 

Thanks Micha, great work 👍 

Merging.
